google/cloud/pubsub/internal/subscription_flow_control_test.cc

```diff
@@ -0,0 +1,379 @@
+      std::string const& prefix, int count) {
+    google::pubsub::v1::PullResponse response;
+    for (int i = 0; i != count; ++i) {
+      auto const id = prefix + std::to_string(i);
+      auto& m = *response.add_received_messages();
+      m.set_ack_id("ack-" + id);
+      m.mutable_message()->set_message_id("message-" + id);
+    }
+    return response;
+  }
+
+  future<void> AddAsyncPull() {
+    std::unique_lock<std::mutex> lk(async_pulls_mu_);
+    promise<void> p;
+    auto f = p.get_future();
+    async_pulls_.push_back(std::move(p));
+    async_pulls_cv_.notify_one();
+    return f;
+  }
+
+  void WaitAsyncPullCount(std::size_t n) {
+    std::unique_lock<std::mutex> lk(async_pulls_mu_);
+    async_pulls_cv_.wait(lk, [&] { return async_pulls_.size() >= n; });
+  }
+
+  std::vector<std::string> CurrentAcks() {
+    std::unique_lock<std::mutex> lk(acks_mu_);
+    auto tmp = std::move(acks_);
+    acks_ = {};
+    return tmp;
+  }
+
+  void SaveAcks(google::pubsub::v1::ReceivedMessage const& m) {
+    std::unique_lock<std::mutex> lk(acks_mu_);
+    acks_.push_back(m.ack_id());
+    lk.unlock();
+    acks_cv_.notify_one();
+  }
+
+  void WaitAcks() {
+    std::unique_lock<std::mutex> lk(acks_mu_);
+    acks_cv_.wait(lk, [&] { return !acks_.empty(); });
+  }
+
+  void WaitAcksCount(std::size_t n) {
+    std::unique_lock<std::mutex> lk(acks_mu_);
+    acks_cv_.wait(lk, [&] { return acks_.size() >= n; });
+  }
+
+  std::mutex async_pulls_mu_;
+  std::condition_variable async_pulls_cv_;
+  std::vector<promise<void>> async_pulls_;
+
+  std::mutex acks_mu_;
+  std::condition_variable acks_cv_;
+  std::vector<std::string> acks_;
+};
+
+TEST_F(SubscriptionFlowControlTest, Basic) {
+  auto mock = std::make_shared<pubsub_testing::MockSubscriptionBatchSource>();
+  EXPECT_CALL(*mock, Shutdown);
+
+  EXPECT_CALL(*mock, Pull)
+      .WillOnce([&](std::int32_t max_messages) {
+        EXPECT_EQ(1, max_messages);
+        return AddAsyncPull().then([](future<void>) {
+          return make_status_or(GenerateMessages("0-", 1));
+        });
+      })
+      .WillOnce([&](std::int32_t max_messages) {
+        EXPECT_EQ(1, max_messages);
+        return AddAsyncPull().then([](future<void>) {
+          return make_status_or(GenerateMessages("1-", 1));
+        });
+      })
+      .WillOnce([&](std::int32_t max_messages) {
+        EXPECT_EQ(1, max_messages);
+        return AddAsyncPull().then([](future<void>) {
+          return make_status_or(GenerateMessages("2-", 1));
+        });
+      });
+  EXPECT_CALL(*mock, AckMessage)
+      .WillOnce([&](std::string const& ack_id, std::size_t) {
+        EXPECT_THAT(ack_id, "ack-0-0");
+        return make_ready_future(Status{});
+      });
+  EXPECT_CALL(*mock, NackMessage)
+      .WillOnce([&](std::string const& ack_id, std::size_t) {
+        EXPECT_THAT(ack_id, "ack-1-0");
+        return make_ready_future(Status{});
+      });
+  // There may be a few of these calls during shutdown, ignore them, we have
```

Comment:
Nit: s/, ignore them, we/, so ignore them. We/ perhaps?

Reply:
Fixed.

Reply:
Or at least "changed", let me know if you think it needs further improvement.